src/OrchardCore.Modules/OrchardCore.AuditTrail/Services/AuditTrailManager.cs

```diff
@@ -0,0 +1,239 @@
+                    EventName = auditTrailCreateContext.EventName,
+                    FullEventName = eventDescriptor.FullEventName,
+                    UserName = !string.IsNullOrEmpty(auditTrailCreateContext.UserName)
+                        ? auditTrailCreateContext.UserName
+                        : T["[empty]"],
+                    CreatedUtc = auditTrailCreateContext.CreatedUtc ?? _clock.UtcNow,
+                    Comment = auditTrailCreateContext.Comment.NewlinesToHtml(),
+                    EventFilterData = auditTrailCreateContext.EventFilterData,
+                    EventFilterKey = auditTrailCreateContext.EventFilterKey,
+                    ClientIpAddress = string.IsNullOrEmpty(auditTrailCreateContext.ClientIpAddress)
+                        ? await GetClientAddressAsync()
+                        : auditTrailCreateContext.ClientIpAddress
+                };
+
+                eventDescriptor.BuildAuditTrailEvent(auditTrailEvent, auditTrailCreateContext.EventData);
+
+                await _auditTrailEventHandlers.InvokeAsync((handler, context, @event)
+                    => handler.AlterAsync(context, @event), auditTrailCreateContext, auditTrailEvent, Logger);
+
+                _session.Save(auditTrailEvent);
+            }
+        }
+
+        public async Task<AuditTrailEventSearchResults> GetAuditTrailEventsAsync(
+            int page,
+            int pageSize,
+            Filters filters = null,
+            AuditTrailOrderBy orderBy = AuditTrailOrderBy.DateDescending)
+        {
+            var query = _session.Query<AuditTrailEvent>();
+
+            if (filters != null)
+            {
+                var filterContext = new QueryFilterContext(query, filters);
+
+                _auditTrailEventHandlers.Invoke((handler, context) => handler.Filter(context), filterContext, Logger);
+
+                // Give each provider a chance to modify the query.
+                var providersContext = DescribeProviders();
+                foreach (var queryFilter in providersContext.QueryFilters)
+                {
+                    queryFilter(filterContext);
+                }
+
+                query = filterContext.Query;
+            }
+
+            switch (orderBy)
+            {
+                case AuditTrailOrderBy.CategoryAscending:
+                    query.With<AuditTrailEventIndex>().OrderBy(eventIndex => eventIndex.Category).ThenByDescending(eventIndex => eventIndex.CreatedUtc);
+                    break;
+                case AuditTrailOrderBy.EventAscending:
+                    query.With<AuditTrailEventIndex>().OrderBy(eventIndex => eventIndex.EventName).ThenByDescending(eventIndex => eventIndex.CreatedUtc);
+                    break;
+                case AuditTrailOrderBy.DateDescending:
+                    query.With<AuditTrailEventIndex>().OrderByDescending(eventIndex => eventIndex.CreatedUtc).ThenByDescending(eventIndex => eventIndex.Id);
```

Comment:
It's actually faster to order by Id (and still accurate) even more if there are lots of items